packages/mobile/src/screens/change-email-screen/ChangeEmailScreen.tsx

```diff
@@ -0,0 +1,184 @@
+import { useCallback, useEffect } from 'react'
+
+import {
+  ChangeEmailPage,
+  useChangeEmailFormConfiguration
+} from '@audius/common/hooks'
+import type {
+  EventListenerCallback,
+  EventMapCore,
+  NavigationState
+} from '@react-navigation/native'
+import { createNativeStackNavigator } from '@react-navigation/native-stack'
+import { Formik, useFormikContext } from 'formik'
+import { TouchableOpacity } from 'react-native'
+
+import { IconArrowRight, IconClose } from '@audius/harmony-native'
+import { BackButton } from 'app/app/navigation/BackButton'
+import {
+  Button,
+  KeyboardAvoidingView,
+  ModalScreen,
+  Screen,
+  ScreenContent
+} from 'app/components/core'
+import { useNavigation } from 'app/hooks/useNavigation'
+import { useToast } from 'app/hooks/useToast'
+import { makeStyles } from 'app/styles'
+
+import { useAppScreenOptions } from '../app-screen/useAppScreenOptions'
+import { VerifyEmailSubScreen } from '../change-password-screen/SubScreens'
+
+import { ConfirmPasswordSubScreen, NewEmailSubScreen } from './SubScreens'
+
+const messages = {
+  change: 'Change Email',
+  continue: 'Continue',
+  success: 'Email updated!'
+}
+
+const useStyles = makeStyles(({ palette, spacing }) => ({
+  screen: {
+    justifyContent: 'space-between'
+  },
+  bottomSection: {
+    overflow: 'hidden',
+    height: 'auto',
+    padding: spacing(4),
+    paddingBottom: spacing(12),
+    backgroundColor: palette.white,
+    borderTopWidth: 1,
+    borderTopColor: palette.neutralLight6
+  }
+}))
+
+const Stack = createNativeStackNavigator()
+
+const ChangeEmailHeaderLeft = ({ page }: { page: ChangeEmailPage }) => {
+  const navigation = useNavigation()
+  if (page === ChangeEmailPage.VerifyEmail) {
+    return <BackButton />
+  } else {
+    return (
+      <TouchableOpacity
+        onPress={() => navigation.navigate('AccountSettingsScreen')}
+      >
+        <IconClose size='l' color='subdued' />
+      </TouchableOpacity>
+    )
+  }
+}
+
+const ChangeEmailNavigator = ({
+  page,
+  setPage
+}: {
+  page: ChangeEmailPage
+  setPage: (page: ChangeEmailPage) => void
+}) => {
+  const { handleSubmit, isSubmitting } = useFormikContext()
+  const styles = useStyles()
+  const navigation = useNavigation()
+
+  // Only show the back button on the OTP page
+  const screenOptions = useAppScreenOptions({
+    headerRight: () => null,
+    headerLeft: () => {
+      return <ChangeEmailHeaderLeft page={page} />
+    }
+  })
+
+  // Map hook page state to screen navigations
+  useEffect(() => {
+    navigation.navigate(ChangeEmailPage[page])
+  }, [page, navigation])
+
+  // Map navigations back to the hook page state
+  const stateListener = useCallback<
```

Comment:
sorry mind explaining the context on this one just a bit more?

Reply:
The navigator uses routes in order to determine which screen to display, but the _page_ from the hook needs to be kept in sync. In order to do that, I set up this listener to the state of the navigator to ensure the page of the hook is aligned with the current screen. That way I can have the various `navigate`s all continue to work fine without having to additionally `setPage` each call site, and trust this event handler will ensure my hook is aware of which page we're on. Likewise, the effect above ensures that the _hook_ tells the _navigator_ where we are, so that if the hook changes pages, the navigator changes screens.

I think this all could have been avoided if each page could have been an independent form. I think my avoidance of redux/a separate context caused this less than ideal circumstance.

Reply:
this is interesting. maybe some sort of higher level concept of a FormStack would be useful